### DISCUSSION TITLE
Will muti-level bracket be supported in rearragement in the future version?

### DISCUSSION THREAD

**Discussion by wzm2256** [arogozhnikov/einops#272] [Q&A] [answered]
Hi again,
while I am doing tensor multiplication, I always need to use bracket rearrangement. (One possible place these brackets appear is kronecker product as showed in my other discussion https://github.com/arogozhnikov/einops/discussions/271)
```
import torch
import einops
A=torch.randn(1, 32,32)

B = einops.rearrange(A, 'a b c -> (a (b c))')

#EinopsError:  Error while processing rearrange-reduction pattern "a b c -> (a (b c))".
#Input tensor shape: torch.Size([32, 32]). Additional info: {}.
#Axis composition is one-level (brackets inside brackets not allowed)

# I have to write it this way
B = einops.rearrange(einops.rearrange(A, 'a b c -> a (b c)'), 'a d -> (a d)')
```
In this example, I have to use two rearrangement twice, and I need to use one more index 'd', which is not intuitive. In practice, the situation can be much more complicated.

**Comment by boeddeker** [comment, marked as answer]
Hi,

you could use a single bracket.
 There is no difference between `(a b c)`, `(a (b c))` and `((a b) c)`.
 Only the letter order is important, not the execution order.
```python
import torch
import einops
A=torch.randn(1, 32,32)
B = einops.rearrange(einops.rearrange(A, 'a b c -> a (b c)'), 'a d -> (a d)')
C = einops.rearrange(A, 'a b c -> (a b c)')
torch.testing.assert_close(B, C)
```

**Reply by wzm2256** [reply]
Oh, I see. The inner bracket can be deleted without infuencing the overall compuatation. That's cool. Thanks!
